ci-tools/test-matrix/src/main.rs

```diff
@@ -0,0 +1,248 @@
+                    let row = suite_map.get_mut(&test_case.name).ok_or_else(|| {
+                        format!("Unknown test-case in junit file: {}", test_case.name)
+                    })?;
+                    let cell = get_at_index_mut(&mut row.cells, run_index)
+                        .as_mut()
+                        .ok_or_else(|| {
+                            format!(
+                                "Unknown test-case for this run in junit file: {}",
+                                test_case.name
+                            )
+                        })?;
+                    cell.status = test_case.status();
+                    cell.output = test_case.output_truncated();
+                    cell.duration = test_case.time;
+                }
+            }
+        }
+        let rows: Vec<_> = row_map
+            .into_iter()
+            .map(|(name, rows)| TestSuite {
+                name,
+                rows: rows.into_values().collect(),
+            })
+            .collect();
+        Ok(Self { rows, columns })
+    }
+}
+
+#[derive(Debug, Default, serde::Serialize)]
+pub struct TestCaseRow {
+    pub name: String,
+    pub cells: Vec<Option<TestCaseCell>>,
+}
+
+#[derive(Copy, Clone, Debug, serde::Serialize)]
+pub enum TestCaseStatus {
+    Passed,
+    Failed,
+    Ignored,
+    Unknown,
+}
+
+#[derive(Clone, Debug, serde::Serialize)]
+pub struct TestCaseCell {
+    pub status: TestCaseStatus,
+    pub output: String,
+    pub duration: f64,
+}
+
+struct TestRun {
+    name: String,
+
+    // The metadata about the tests to run (includes tests that were ignored)
+    test_list: TestListSummary,
+
+    // The results of the tests that were run (doesn't include ignored tests)
+    junit_suites: junit::TestSuites,
+}
+impl TestRun {
+    fn from_zip_bytes(name: String, bytes: &[u8]) -> Result<TestRun, Box<dyn Error>> {
+        let mut archive = zip::ZipArchive::new(Cursor::new(bytes))?;
+        let json = zip_extract_file(&mut archive, "nextest-list.json")?;
+        let json = String::from_utf8(json)?;
+        let test_list = nextest_metadata::TestListSummary::parse_json(json)?;
+        let junit_xml = String::from_utf8(zip_extract_file(&mut archive, "junit.xml")?)?;
+        let junit_suites = junit::TestSuites::from_xml(&junit_xml)?;
+
+        Ok(TestRun {
+            name,
+            test_list,
+            junit_suites,
+        })
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Serialize)]
+pub struct RunInfo {
+    pub id: String,
+    pub display_name: String,
+}
+impl RunInfo {
+    fn from_run(run: &Run) -> Self {
+        RunInfo {
+            id: run.created_at.format("%F-%H%M%S").to_string(),
+            display_name: run.created_at.format("%F %H:%M:%S").to_string(),
+        }
+    }
+}
+
+#[tokio::main(flavor = "current_thread")]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let www_out = std::env::var("CPTRA_WWW_OUT").unwrap();
```

Comment:
If these aren't present, we should print a message what what the caller should do to populate them rather than panicking.

Reply:
Added expectation message. FWIW, this isn't really intended to be used by humans.